src/cider/nrepl/middleware/inspect.clj

```diff
@@ -47,77 +48,35 @@
   [handler msg]
   (handler (eval-msg msg)))
 
-(defn pop-reply
-  [{:keys [transport] :as msg}]
-  (try
-    (let [inspector (swap-inspector! msg inspect/up)]
-      (transport/send
-       transport
-       (response-for msg :value (:rendered inspector) :status :done)))
-    (catch Exception e
-      (transport/send
-       transport
-       (response-for msg (u/err-info e :inspect-pop-error))))))
-
-(defn push-reply
-  [{:keys [idx transport] :as msg}]
-  (try
-    (let [inspector (swap-inspector! msg inspect/down idx)]
-      (transport/send
-       transport
-       (response-for msg :value (:rendered inspector) :status :done)))
-    (catch Exception e
-      (transport/send
-       transport
-       (response-for msg (u/err-info e :inspect-push-error))))))
-
-(defn refresh-reply
-  [{:keys [transport] :as msg}]
-  (try
-    (let [inspector (swap-inspector! msg #(or % (inspect/fresh)))]
-      (transport/send
-       transport
-       (response-for msg :value (:rendered inspector) :status :done)))
-    (catch Exception e
-      (transport/send
-       transport
-       (response-for msg (u/err-info e :inspect-refresh-error))))))
-
-(defn next-page-reply
-  [{:keys [transport] :as msg}]
-  (try
-    (let [inspector (swap-inspector! msg inspect/next-page)]
-      (transport/send
-       transport
-       (response-for msg :value (:rendered inspector) :status :done)))
-    (catch Exception e
-      (transport/send
-       transport
-       (response-for msg (u/err-info e :inspect-next-page-error))))))
-
-(defn prev-page-reply
-  [{:keys [transport] :as msg}]
-  (try
-    (let [inspector (swap-inspector! msg inspect/prev-page)]
-      (transport/send
-       transport
-       (response-for msg :value (:rendered inspector) :status :done)))
-    (catch Exception e
-      (transport/send
-       transport
-       (response-for msg (u/err-info e :inspect-prev-page-error))))))
-
-(defn set-page-size-reply
-  [{:keys [page-size transport] :as msg}]
-  (try
-    (let [inspector (swap-inspector! msg inspect/set-page-size page-size)]
-      (transport/send
-       transport
-       (response-for msg :value (:rendered inspector) :status :done)))
-    (catch Exception e
-      (transport/send
-       transport
-       (response-for msg (u/err-info e :inspect-set-page-size-error))))))
+(defn- success [{:keys [transport] :as msg} inspector]
+  (transport/send transport (response-for msg :value (:rendered inspector) :status :done)))
+
+(defn- failure [{:keys [transport] :as msg} err err-kw]
+  (transport/send transport (base-error-response msg err err-kw :done)))
+
+(defn pop-reply [msg]
+  (try (success msg (swap-inspector! msg inspect/up))
+       (catch Exception e (failure msg e :inspect-pop-error))))
```

Comment:
You're right, the Elisp code has **absolutely no idea** about these error codes. They could be removed entirely, but I think they're good to have so that we can "tune" the error reporting UI based on the type of error returned (which is something I've been experimenting with). Eval/big errors should certainly get the stacktrace viewer, some other types should just get a minibuffer message, and other types of errors might just get swallowed. The user would also be able to customize this by fooling around with some variables.

I think there are definitely a lot of places to clean up the op handling code across the board. The biggest gain would probably involve using the "inline-response map" feature of the macro and getting rid of all these `<op>-reply` functions that now basically don't do anything. I think this would really give us a clear line between interface and implementation as well.